### PR TITLE
Physics by bullet added.

### PR DESCRIPTION
**Pull request by Schtekt**
Collision detection/response added to the engine.

Collision components may be added to entities that have transform components. Make sure that the shape of the collision component matches the object ;D

### CODE REVIEW

File: StortSpelProjekt/Engine/src/ECS/Components/Collision/CollisionComponent.cpp [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+void component::CollisionComponent::InitScene()
+{
+#ifdef _DEBUG
+	// If no transform is given, this component is useless!
+	if (!m_pParent->HasComponent<component::TransformComponent>())
+	{
+		Log::PrintSeverity(Log::Severity::CRITICAL, "No Transform provided for collisioncomponent in entity %s\n", m_pParent->GetName().c_str());
+		return;
+	}
+#endif
```
Comment (AntonAsbrink) [inline StortSpelProjekt/Engine/src/ECS/Components/Collision/CollisionComponent.cpp]: This does not run every frame and should therefore still exist in release for error checking purposes

File: StortSpelProjekt/Engine/src/Misc/Window.cpp [hunk 2]
```diff
@@ -181,7 +181,7 @@ bool Window::ExitWindow()
 	{
 
 		TranslateMessage(&msg);
-		DispatchMessage(&msg);	// G�r in i "CALLBACK" funktionen
+		DispatchMessage(&msg);	// G�r in i "CALLBACK" funktionen
```
Comment (AntonAsbrink) [inline StortSpelProjekt/Engine/src/Misc/Window.cpp]: No swedish

File: StortSpelProjekt/Engine/src/Physics/Physics.cpp [hunk 3]
```diff
@@ -65,5 +108,20 @@ void Physics::collisionChecks(double dt)
...
+	//m_timeSinceLastColCheck += dt;
+
+	for (int i = 0; i < m_CollisionComponents.size(); i++)
+	{
+		for (int j = i + 1; j < m_CollisionComponents.size(); j++)
+		{
+			//m_CollisionComponents[i]->CheckCollision(m_CollisionComponents[j]);
+		}
 	}
 }
```
Comment (AntonAsbrink) [inline StortSpelProjekt/Engine/src/Physics/Physics.cpp]: obviously deprecated function should be removed

File: StortSpelProjekt/Engine/src/Physics/Physics.h [hunk 4]
```diff
@@ -20,23 +35,32 @@ class Physics
...
-
+	void AddCollisionComponent(component::CollisionComponent* comp);
+	void RemoveCollisionComponent(component::CollisionComponent* comp);
+	const btDynamicsWorld* GetWorld();
 private:
 	Physics();
 	double m_timeSinceLastColCheck = 0;
 	// How often the collisions are checked
-	const double m_CollisionUpdateInterval = 0.2;
+	const double m_CollisionUpdateInterval = 0.01;
```
Comment (AntonAsbrink) [inline StortSpelProjekt/Engine/src/Physics/Physics.h]: Does not follow code standard.
Also default values for variables should be in constructor and not header.

File: StortSpelProjekt/Engine/src/Physics/Physics.h [hunk 5]
```diff
@@ -20,23 +35,32 @@ class Physics
...
+	btCollisionConfiguration* m_pCollisionConfig;
+	btDbvtBroadphase* m_pBroadphase;
+	btConstraintSolver* m_pSolver;
 
 	// Checks collision for all entities in the collison entities vector
 	// publishes an event if a collision has happened
 	void collisionChecks(double dt);
+	// Checks collision for all added collision components
+	void collisionComponentChecks();
 	// prints which entities that have collieded to the consol
```
Comment (AntonAsbrink) [inline StortSpelProjekt/Engine/src/Physics/Physics.h]: Spelling error *collided *console

File: StortSpelProjekt/Game/src/Gamefiles/Components/PlayerInputComponent.cpp [hunk 6]
```diff
@@ -36,20 +33,46 @@ void component::PlayerInputComponent::Init()
...
+	if (!m_pCamera)
+	{
+		Log::PrintSeverity(Log::Severity::CRITICAL, "PlayerInputComponent needs a Camera component!\n");
+	}
+
+	if (!m_pTransform)
+	{
+		Log::PrintSeverity(Log::Severity::CRITICAL, "PlayerInputComponent needs a Transform component!\n");
+	}
+#endif
```
Comment (AntonAsbrink) [inline StortSpelProjekt/Game/src/Gamefiles/Components/PlayerInputComponent.cpp]: Init functions should still run error testing

**Review by AntonAsbrink — CHANGES_REQUESTED** [review]

**Review by AntonAsbrink — DISMISSED** [review]

**Review by kattalainen — APPROVED** [review]

**Review by AntonAsbrink — APPROVED** [review]

**Pull request merged**